presto-main/src/main/java/com/facebook/presto/catalogserver/RemoteMetadataManager.java

```diff
@@ -0,0 +1,778 @@
+        this.transactionManager = requireNonNull(transactionManager, "transactionManager is null");
+        this.objectMapper = requireNonNull(objectMapper, "objectMapper is null");
+        this.catalogServerClient = requireNonNull(catalogServerClient, "catalogServerClient is null");
+    }
+
+    @Override
+    public final void verifyComparableOrderableContract()
+    {
+        delegate.verifyComparableOrderableContract();
+    }
+
+    @Override
+    public Type getType(TypeSignature signature)
+    {
+        return delegate.getType(signature);
+    }
+
+    @Override
+    public void registerBuiltInFunctions(List<? extends SqlFunction> functionInfos)
+    {
+        delegate.registerBuiltInFunctions(functionInfos);
+    }
+
+    @Override
+    public boolean schemaExists(Session session, CatalogSchemaName schema)
+    {
+        return catalogServerClient.get().schemaExists(
+                transactionManager.getTransactionInfo(session.getRequiredTransactionId()),
+                session.toSessionRepresentation(),
+                schema);
+    }
+
+    @Override
+    public boolean catalogExists(Session session, String catalogName)
+    {
+        return catalogServerClient.get().catalogExists(
+                transactionManager.getTransactionInfo(session.getRequiredTransactionId()),
+                session.toSessionRepresentation(),
+                catalogName);
+    }
+
+    @Override
+    public List<String> listSchemaNames(Session session, String catalogName)
+    {
+        String schemaNamesJson = catalogServerClient.get().listSchemaNames(
+                transactionManager.getTransactionInfo(session.getRequiredTransactionId()),
+                session.toSessionRepresentation(),
+                catalogName);
+        if (!schemaNamesJson.equals(EMPTY_STRING)) {
+            return readValue(schemaNamesJson, new TypeReference<List<String>>() {});
+        }
+        return new ArrayList<>();
+    }
+
+    @Override
+    public Optional<TableHandle> getTableHandle(Session session, QualifiedObjectName table)
+    {
+        String tableHandleJson = catalogServerClient.get().getTableHandle(
+                transactionManager.getTransactionInfo(session.getRequiredTransactionId()),
+                session.toSessionRepresentation(),
+                table);
+        if (!tableHandleJson.equals(EMPTY_STRING)) {
+            TableHandle tableHandle = readValue(tableHandleJson, new TypeReference<TableHandle>() {});
+            Optional<CatalogMetadata> catalogMetadata = this.transactionManager.getOptionalCatalogMetadata(session.getRequiredTransactionId(), table.getCatalogName());
+            if (catalogMetadata.isPresent()) {
+                ConnectorTransactionHandle connectorTransactionHandle = catalogMetadata.get().getTransactionHandleFor(tableHandle.getConnectorId());
+                tableHandle.setTransaction(connectorTransactionHandle);
+            }
```

Comment:
It think either way is fine however, im not sure how much overhead recreating the tableHandle would cause. I think making it mutable should be fine unless there are some drawbacks do doing so in the codebase then we should probably recreate the tableHandle instead

Reply:
Was recreated for now